### PR TITLE
Document extra-packages a bit better and clean up related entries (closes #5448)

### LINKED ISSUE #5448: Document `extra-packages` in user's guide
see https://github.com/haskell/cabal/issues/5444#issuecomment-406195007 for pointers

### PR DESCRIPTION
**Pull request by linearray**
Please review carefully, I have never used any of these and only documented them in the way it makes sense to me, which is not necessarily the correct one.

closes #5448 

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: doc/nix-local-build.rst [hunk 1]
```diff
@@ -146,11 +146,11 @@ must be built per-project, versus external packages, which can be cached
 across projects. To be more precise:
 
 1. A **local package** is one that is listed explicitly in the
-   ``packages``, ``optional-packages`` or ``extra-packages`` field of a
-   project. Usually, these refer to packages whose source code lives
-   directly in a folder in your project. But you can list an
-   arbitrary Hackage package in :cfg-field:`packages`
-   to force it to be treated as local.
+   ``packages``, ``optional-packages`` and ``extra-packages`` fields of a
```
Comment (ffaf1) [inline doc/nix-local-build.rst:149]: “or” or “and”? If it is “in one of those fields”, I would say “or”.

File: doc/nix-local-build.rst [hunk 2]
```diff
@@ -146,11 +146,11 @@ must be built per-project, versus external packages, which can be cached
...
+   project. They refer to packages, whose source code usually lives
```
Comment (ffaf1) [inline doc/nix-local-build.rst:150]: Not a native speaker but I feel the comma is unnecessary

File: doc/cabal-project.rst [hunk 3]
```diff
@@ -127,8 +127,10 @@ project are:
 .. cfg-field:: extra-packages: package list with version bounds (comma separated)
     :synopsis: Adds external packages as local
 
-    :strike:`Specifies a list of external packages from Hackage which
-    should be considered local packages.` (Not implemented)
+    Specifies a list of external packages from Hackage, which
+    should be considered local packages. The motivation for
+    :cfg-field:`extra-packages` is making libraries that are not
```
Comment (ffaf1) [inline doc/cabal-project.rst:132]: Is this one possible motivation or *the* motivation?

Comment (linearray) [reply doc/cabal-project.rst:132]: Taken straight from the horse's mouth:
https://github.com/haskell/cabal/issues/3737#issuecomment-243718434

I imagine there are other uses, but I don't know. 

"A common use case" would be an alternative.

File: doc/nix-local-build.rst (hunk 2 above)
Comment (linearray) [reply doc/nix-local-build.rst]: I rewrote the sentence, because it was too complicated anyway.

File: doc/cabal-project.rst (hunk 3 above)
Comment (linearray) [reply doc/cabal-project.rst:132]: Just found https://github.com/haskell/cabal/issues/6952 and the related https://github.com/haskell/cabal/issues/7865 and it seems you can install extra executables with ``extra-packages`` too, but there are caveats. Not sure if it should go into the documentation, while those are unresolved.

**Review by ffaf1 — APPROVED** [review]
good job!

**Review by jneira — APPROVED** [review]
lgtm, thanks for improving docs

**Comment by Mikolaj** [comment]
@linearray: thanks a lot for the docs improvement. BTW, we normally use the `merge_me` label to merge PRs:

https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions

Please give it a try next time, it's pretty handy. :)

**Pull request merged**